src/main/java/org/sqlite/core/DB.java

```diff
@@ -80,6 +80,16 @@ public SQLiteConfig getConfig() {
 
     // WRAPPER FUNCTIONS ////////////////////////////////////////////
 
+    /**
+     * Determines whether the database named databaseName of this connection is read-only.
+     *
+     * @param databaseName name of the database to inquire about (may be null)
+     * @throws SQLException if the connection is already closed
+     * @see <a href="https://www.sqlite.org/c3ref/db_readonly.html">https://www.sqlite.org/c3ref/db_readonly.html</a>
+     *      <a href="https://www.sqlite.org/c3ref/open.html">https://www.sqlite.org/c3ref/open.html</a>
+     */
+    public abstract boolean isReadOnly(final String databaseName) throws SQLException;
```

Comment:
what 's the point of passing the database name ? is that to handle attached databases?

returning a boolean doesn't handle the case where the provided db name is not available on the connection:

> The sqlite3_db_readonly(D,N) interface returns 1 if the database N of connection D is read-only, 0 if it is read/write, or -1 if N is not the name of a database on connection D.

Reply:
yes, to allow for checking against attached databases. there's a bit of a disconnect between the traditional concept of a connection (that only connects to a single database) and sqlite's implementation of a connection as a collection of traditional connections.

the native method does allow for differentiating all the conditions. `isReadOnly` abstracts away the case of the not found database into "well, it's not read only". this main user of this method will be the implementer of Connection.isReadOnly, and their only great option is to pass "main" (which always exists in sqlite), so pushing a more than boolean result back up to that caller doesn't provide much value that i can see. i could return an enum out of this isReadOnly, but i don't see that actually being useful, especially when we could add another method to specifically test whether a database is found on the connection just by calling is_read_only and checking for -1.

Reply:
I feel like the DB implementation should mirror what the sqlite function returns, then we can abstract what we want to return in user land. We may want to throw an exception if a named database doesn't exist on a connection for instance.

As for `java.sql.Connection#isReadOnly`, which is not yet in this PR (and should be IMO), it probably should consider all the attached databases in order to say whether the connection is read only or not, don't you think?

Reply:
that seems like a slippery slope into using exceptions for flow control. referring to conventional wisdom: "Item 69: Use exceptions only for exceptional conditions" in [effective java 3rd ed](https://github.com/GunterMueller/Books-3/raw/master/Effective%20Java%20(3rd%20Edition).pdf)

`A class with a "state-dependent" method that can be invoked only under certain unpredictable conditions should generally have a separate "state-testing" method indicating whether it is appropriate to invoke the state-dependent method.`

with that in mind, how about we add an abstract method to DB.java and then implement it in NativeDB.java like this:
```
    @Override
    public boolean hasDatabase(final String databaseName) throws SQLException {
        return this.is_read_only(databaseName) != -1;
    }
```
with this, a caller can check and determine if the database is available on the connection.

it's just a matter of style. if you'd rather, i can make DB.isReadOnly return an enum with the 3 states of readonly, not readonly, and not found. maybe you prefer this as it avoids one native call. i could go either way as long as we don't venture into using exceptions for flow control.

regarding `java.sql.Connection#isReadOnly`, i don't think we should consider all databases available on the connection by default. some databases can be read/write while others are read-only. the temp database, and memory databases, i think, are always read-write, so we could not do an AND between all databases.
we could return that we are read-only if any of the databases are read-only.
maybe we make a SQLiteConfig option to let people choose how to determine if the connection is considered read-only. that's probably the best way to avoid tripping some people up. we'd default to basing the decision on only the "main" database, like it does now.

for reference, we can get a list of databases currently on the connection with:
`pragma database_list;` or `select * from pragma_database_list;`

Reply:
I am not saying we should use exceptions for control flow, just that exposing more information in the DB function seems like a good idea.

> i don't think we should consider all databases available on the connection by default. some databases can be read/write while others are read-only

Exactly. But JDBC doesn't care about how many DBs are on the connection, it needs to know whether it is read only or not.

> so we could not do an AND between all databases.
> we could return that we are read-only if any of the databases are read-only

the end result is the same. If any DB on the connection is read only, it should return `true`

> maybe we make a SQLiteConfig option to let people choose how to determine if the connection is considered read-only

I am not in favor of that. JDBC always needs some interpretation anyway, depending on the underlying database.



> we'd default to basing the decision on only the "main" database, like it does now.

at the moment we only check the configuration AFAIK, and it would apply to attached DB too IIUC.

Reply:
> I am not in favor of that. JDBC always needs some interpretation anyway, depending on the underlying database.

what i'm advancing is letting the user choose the interpretation by telling the driver via a config option sent into the java.sql.Connection. so, there'd be a config option that java.sql.Connection.isReadOnly uses to determine if it should return true based on a) only the read-only status of the "main" database, or b) the read-only status of any database. i'm excluding the possibility of the result being determined by the status of all databases because when you add a temp database, or a memory database, i'm pretty sure that you'd be locking into never being able to return read-only.

i'm trying to support people that connect initially to a read-write main database and then attach to another read-only database. i can think of valid uses of this sequence of events.

Reply:
i don't think we need that kind of knob tbh

Reply:
i was thinking of a scenario where someone had a read+write main database. then they attach to a read-only database in the same connection that contains info that they use to make updates to the main database. it's possible that people have existing code doing that, and it wouldn't be crazy. to open the attached database in read only mode, i believe all that has to be done is to make the file read-only with file permissions (or maybe even just mount the filesystem containing the second database as read-only).

so, the existing code would consider the connection read-write, but if we start considering the connection read-only if any one database is read-only, then it would then become read-only, which would be a breaking change. therefore, i think we'd default to the old way of considering the connection read-only depending on only the status of the main database. then we could give people the knob that allows them to tell the SQLiteConnection.isReadOnly implementation whether to consider the connection read-only if any of the additional databases are read-only.

Reply:
just pushed up a commit that returns the database status as an enum to match the native call.